xc/device.ml

```diff
@@ -1414,8 +1426,71 @@ module PCI = struct
       Forkhelpers.execute_command_get_output !Resources.rmmod ["i915"] in ()
 
   let procfs_nvidia = "/proc/driver/nvidia/gpus"
+  let nvidia_smi    = "/usr/bin/nvidia-smi"
+  let nvidia_manage = "/usr/lib/nvidia/sriov-manage"
+
+  (** [num_vfs devstr] returns the number of PCI VFs of [devstr] or 0 if
+    * [devstr] is not an SRIOV device
+    *)
+  let num_vfs devstr =
+    let path = Printf.sprintf "/sys/bus/pci/devices/%s/sriov_numvfs" devstr in
+    try
+      Unixext.string_of_file path |> String.trim |> int_of_string
+    with
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+        debug "File %s does not exist - assuming no SRIOV devices in use" path;
+        0
+      | exn ->
+        let msg = Printf.sprintf "Can't read %s to reset Nvidia GPU %s: %s"
+          path devstr (Printexc.to_string exn) in
+        error "%s" msg;
+        raise (Xenopsd_error (Internal_error msg))
+
+  (** [vfs_of device] returns the PCI addresses of the virtual functions
+   * of PCI [device]. We find each virtual function by looking at the
+   * virtfnX symlink in [device].
+   *)
+  let vfs_of devstr =
+    let device = Filename.concat sysfs_devices devstr in
+    let some x = Some x in
+    let virtfn n =
+      let path = Printf.sprintf "virtfn%d" n in
+      try
+        Filename.concat device path |> Unix.readlink |> Filename.basename |> some
+      with exn ->
+        debug "Can't read %s of PCI %s: %s" path devstr (Printexc.to_string exn);
+        None in
+    let rec seq = function
+      | n when n <= 0 -> []
+      | n -> n-1 :: seq (n-1)
+    in
+    num_vfs devstr |> seq |> List.filter_map virtfn
 
-  let nvidia_smi = "/usr/bin/nvidia-smi"
+  (** [deactivate_nvidia_sriov devstr] deactivates SRIOV PCI VFs of [devstr] if
+    * necessary. This needs to be called for NVidia GPUs before using
+    * [devstr] as a pass-through GPU.
+    *)
+  let deactivate_nvidia_sriov devstr =
+    let cmd  = nvidia_manage in
+    let args = ["-d"; devstr] in
+
+    match vfs_of devstr with
+    | [] ->
+      debug "No need to deactivate NVidia vGPUs %s (it has 0 VFs)" devstr
+    | vfs when Sys.file_exists cmd ->
+      vfs |> List.iter unbind_from_pciback;
+      debug "About to deactivate NVidia vGPUs %s by calling %s" devstr cmd;
+      let out, _ = Forkhelpers.execute_command_get_output cmd args in
```

Comment:
Looks like this is "best effort"? We are not checking the result of this command?

Reply:
This raises an exception if the commend fails - as far as I read the code of `Forkhelpers.execute_command_get_output`